dev/src/transaction.ts

```diff
@@ -187,8 +191,29 @@ export class Transaction implements firestore.Transaction {
 
     const documentReader = new DocumentReader(this._firestore, documents);
     documentReader.fieldMask = fieldMask || undefined;
-    documentReader.transactionId = this._transactionId;
-    return documentReader.get(this._requestTag);
+
+    if (!this._optimisticLocking) {
+      documentReader.transactionId = this._transactionId;
+    }
+
+    return documentReader.get(this._requestTag).then(docs => {
+      for (const doc of docs) {
+        if (doc.exists) {
+          this._locks.set(
+            doc.ref.formattedName,
+            new Precondition({
+              lastUpdateTime: doc.updateTime,
+            })
+          );
+        } else {
+          this._locks.set(
```

Comment:
What's the expected behavior if an optimistic transaction reads the same document twice, but gets a different value each time?

My intuition says the tx should fail since the document value changed within the scope of the transaction, and if that's the intended behavior, we should check that the document isn't in the map.